File: app/src/main/java/io/github/anenasa/news/Channel.java
package io.github.anenasa.news;

import com.chaquo.python.PyException;
import com.chaquo.python.PyObject;
import com.chaquo.python.Python;

import org.json.JSONException;
import org.json.JSONObject;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;

public class Channel {
    public static final int NEEDPARSE_NO = 0;
    public static final int NEEDPARSE_YES = 1;
    public static final int NEEDPARSE_UNKNOWN = 2;

    final String defaultUrl;
    final String defaultName;
    final String defaultFormat;
    final float defaultVolume;
    final String defaultHeader;
    final Map<String, String> ytdlOptions;
    String video = "";
    String customUrl = "";
    String customName = "";
    String customFormat = "";
    String customVolume = "";
    String customHeader = "";
    boolean hidden = false;

    public Channel(String url, String name, String format, float volume, String header, Map<String, String> ytdlOptions) {
        this.defaultUrl = url;
        this.defaultName = name;
        this.defaultFormat = format;
        this.defaultVolume = volume;
        this.defaultHeader = header;
        this.ytdlOptions = ytdlOptions;
    }

    public String getUrl(){
        if(customUrl.isEmpty()){
            return defaultUrl;
        }
        else{
            return customUrl;
        }
    }

    public String getName(){
        if(customName.isEmpty()) {
            return defaultName;
        }
        else{
            return customName;
        }
    }

    public String getFormat(){
        if(customFormat.isEmpty()) {
            return defaultFormat;
        }
        else{
            return customFormat;
        }
    }

    public float getVolume(){
        if(customVolume.isEmpty()) {
            return defaultVolume;
        }
        else{
            return Float.parseFloat(customVolume);
        }
    }

    public String getHeader(){
        if(customHeader.isEmpty()) {
            return defaultHeader;
        }
        else{
            return customHeader;
        }
    }

    public Map<String, String> getHeaderMap(){
        String[] headers = getHeader().split("\\\\r\\\\n");
        Map<String, String> map = new HashMap<>();
        for (String header : headers) {
            int pos = header.indexOf(":");
            if(pos != -1){
                map.put(header.substring(0, pos), header.substring(pos + 1));
            }
        }
        return map;
    }

    public String getVideo(){
        return video;
    }

    public boolean isHidden(){
        return hidden;
    }

    public void setVideo(String video) {
        this.video = video;
    }

    public void setUrl(String url){
        customUrl = url;
    }

    public void setName(String name){
        customName = name;
    }

    public void setFormat(String format){
        customFormat = format;
    }

    public void setVolume(String volume){
        customVolume = volume;
    }

    public void setHeader(String header){
        customHeader = header;
    }

    public void setHidden(boolean hidden){
        this.hidden = hidden;
    }

    int needParse(){
        String url = getUrl();
        if(getVideo().isEmpty() || url.startsWith("https://today.line.me")) {
            return NEEDPARSE_YES;
        }
        if(url.endsWith("m3u8")){
            return NEEDPARSE_NO;
        }
        if(url.startsWith("https://www.youtube.com/")){
            long current = System.currentTimeMillis() / 1000;
            int pos = getVideo().indexOf("expire") + 7;
            long expire = Long.parseLong(getVideo().substring(pos, pos + 10));
            if(current < expire){
                return NEEDPARSE_NO;
            }
            else{
                return NEEDPARSE_YES;
            }
        }
        if(url.startsWith("https://hamivideo.hinet.net/") || url.startsWith("https://embed.4gtv.tv/") ||
                url.startsWith("https://www.ftvnews.com.tw/live/live-video/1/")){
            long current = System.currentTimeMillis() / 1000;
            int pos = getVideo().indexOf("expires") + 8;
            long expire = Long.parseLong(getVideo().substring(pos, pos + 10));
            if(current < expire){
                return NEEDPARSE_NO;
            }
            else{
                return NEEDPARSE_YES;
            }
        }
        return NEEDPARSE_UNKNOWN;
    }

    void parse(YtDlp ytdlp) throws JSONException, IOException, InterruptedException, PyException, InvalidAlgorithmParameterException, IllegalBlockSizeException, NoSuchPaddingException, BadPaddingException, NoSuchAlgorithmException, InvalidKeyException {
        String url = getUrl();
        if(url.startsWith("https://hamivideo.hinet.net/") && url.endsWith(".do")){
            String id = url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf("."));
            OkHttpClient okHttpClient = new OkHttpClient();
            Request.Builder okHttpRequestBuilder = new Request.Builder()
                    .url("https://hamivideo.hinet.net/api/play.do?freeProduct=1&id=" + id);
            for(Map.Entry<String, String> entry : getHeaderMap().entrySet()){
                okHttpRequestBuilder.addHeader(entry.getKey(), entry.getValue());
            }
            Request okHttpRequest = okHttpRequestBuilder.build();
            Response response = okHttpClient.newCall(okHttpRequest).execute();
            ResponseBody body = response.body();
            if (body == null) throw new IOException("body is null");
            JSONObject object = new JSONObject(body.string());
            url = object.getString("url");
        }
        else if(url.equals("https://news.ebc.net.tw/live")){
            Document doc = Jsoup.connect(url).get();
            Element el = doc.selectFirst("div#live-slider div.live-else-little-box");
            if(el == null) throw new IOException("找不到 div");
            url = el.attr("data-code");
        }
        else if(url.startsWith("https://today.line.me/tw/v2/article/")){
            Document doc = Jsoup.connect(url).get();
            Element el =doc.selectFirst("script:containsData(__NUXT__)");
            if(el == null) throw new IOException("找不到 script");
            String script = el.data();
            String id = script.substring(script.indexOf("broadcastId:")+13);
            id = id.substring(0,id.indexOf("\""));
            OkHttpClient okHttpClient = new OkHttpClient();
            Request okHttpRequest = new Request.Builder()
                    .url("https://today.line.me/webapi/glplive/broadcasts/" + id)
                    .build();
            Response response = okHttpClient.newCall(okHttpRequest).execute();
            ResponseBody body = response.body();
            if (body == null) throw new IOException("body is null");
            JSONObject object = new JSONObject(body.string());
            url = object.getJSONObject("hlsUrls").getString("abr");
        }
        else if(url.startsWith("https://embed.4gtv.tv/") || url.startsWith("https://www.ftvnews.com.tw/live/live-video/1/")){
            String id;
            if(url.startsWith("https://www.ftvnews.com.tw/live/live-video/1/")){
                id = url.substring(url.lastIndexOf("/") + 1);
            }
            else{
                Document doc = Jsoup.connect(url).get();
                Element el = doc.selectFirst("script:containsData(ChannelId)");
                if(el == null) throw new IOException("找不到 script");
                String script = el.data();
                id = script.substring(script.indexOf("ChannelId") + 12);
                id = id.substring(0, id.indexOf("\""));
            }
            OkHttpClient okHttpClient = new OkHttpClient();
            Request okHttpRequest = new Request.Builder()
                    .url("https://app.4gtv.tv/Data/GetChannelURL_Mozai.ashx?callback=channelname&Type=LIVE&ChannelId=" + id)
                    .build();
            Response response = okHttpClient.newCall(okHttpRequest).execute();
            ResponseBody body = response.body();
            if (body == null) throw new IOException("body is null");
            String bodyString = body.string();
            String videoUrl = bodyString.substring(bodyString.indexOf("VideoURL") + 11);
            url = videoUrl.substring(0, videoUrl.indexOf("\""));
        }
        else if(url.startsWith("https://www.litv.tv/channel/watch.do")){
            String id = url.substring(url.indexOf("content_id") + 11);
            if(id.contains("&")) id = id.substring(0, id.indexOf("&"));

            OkHttpClient okHttpClient = new OkHttpClient();
            String data = String.format("{\"type\":\"auth\",\"contentId\":\"%s\",\"contentType\":\"channel\"}", id);
            RequestBody requestBody = RequestBody.create(data, MediaType.parse("application/json"));
            Request.Builder okHttpRequestBuilder = new Request.Builder()
                    .url("https://www.litv.tv/channel/ajax/getUrl")
                    .post(requestBody);
            for(Map.Entry<String, String> entry : getHeaderMap().entrySet()){
                okHttpRequestBuilder.addHeader(entry.getKey(), entry.getValue());
            }
            Request okHttpRequest = okHttpRequestBuilder.build();
            Response response = okHttpClient.newCall(okHttpRequest).execute();
            ResponseBody body = response.body();
            if (body == null) throw new IOException("body is null");
            JSONObject object = new JSONObject(body.string());
            url = object.getString("fullpath");
        }
        else if(url.startsWith("https://www.ofiii.com/channel/watch/")){
            OkHttpClient okHttpClient = new OkHttpClient();
            // Get device id
            Request deviceidRequest = new Request.Builder().url("https://www.ofiii.com/api/deviceId").build();
            Response deviceidResponse = okHttpClient.newCall(deviceidRequest).execute();
            ResponseBody deviceidBody = deviceidResponse.body();
            if (deviceidBody == null) throw new IOException("body is null");
            String deviceid = deviceidBody.string().replace("\"","");

            String id = url.substring(url.lastIndexOf("/") + 1);
            Request.Builder okHttpRequestBuilder = new Request.Builder()
                    .url(String.format("https://cdi.ofiii.com/ofiii_cdi/video/urls?device_type=pc&device_id=%s&media_type=channel&asset_id=%s&project_num=OFWEB00", deviceid, id));
            for(Map.Entry<String, String> entry : getHeaderMap().entrySet()){
                okHttpRequestBuilder.addHeader(entry.getKey(), entry.getValue());
            }
            Request okHttpRequest = okHttpRequestBuilder.build();
            Response response = okHttpClient.newCall(okHttpRequest).execute();
            ResponseBody body = response.body();
            if (body == null) throw new IOException("body is null");
            JSONObject object = new JSONObject(body.string());
            url = object.getJSONArray("asset_urls").getString(0);
        }

        PyObject option = Python.getInstance().getBuiltins().callAttr("dict");
        option.callAttr("__setitem__", "format", getFormat());
        if(!getHeader().isEmpty()) {
            PyObject header_dict = Python.getInstance().getBuiltins().callAttr("dict");
            for(Map.Entry<String, String> entry : getHeaderMap().entrySet()){
                header_dict.callAttr("__setitem__", entry.getKey(), entry.getValue());
            }
            option.callAttr("__setitem__", "http_headers", header_dict);
        }
        for (Map.Entry<String, String> entry : this.ytdlOptions.entrySet()) {
            option.callAttr("__setitem__", entry.getKey(), entry.getValue());
        }
        setVideo(ytdlp.extract(url, option));
    }
}

File: app/src/main/java/io/github/anenasa/news/Channel.kt
package io.github.anenasa.news

import android.util.Log
import com.chaquo.python.Python
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import org.jsoup.Jsoup
import java.io.IOException

class Channel(
    val defaultUrl: String,
    val defaultName: String,
    val defaultFormat: String,
    val defaultVolume: Float,
    val defaultHeader: String,
    val ydlOptions: MutableMap<String, String>
) {
    @Volatile
    var video: String = ""
    @Volatile
    var customUrl: String = ""
    @Volatile
    var customName: String = ""
    @Volatile
    var customFormat: String = ""
    @Volatile
    var customVolume: String = ""
    @Volatile
    var customHeader: String = ""
    @Volatile
    var isHidden: Boolean = false
    val headerMap: MutableMap<String, String>
    @Volatile
    var time: Long = 0L
    private val mutex = Mutex()

    init {
        headerMap = header.split("\\r\\n")
            .filter { it.isNotEmpty() }
            .associate {
                it.substringBefore(":") to it.substringAfter(":").removePrefix(" ")
            }
            .toMutableMap()
    }

    var url: String
        get() {
            return customUrl.ifEmpty { defaultUrl }
        }
        set(url) {
            customUrl = url
        }

    var name: String
        get() {
            return customName.ifEmpty { defaultName }
        }
        set(name) {
            customName = name
        }

    var format: String
        get() {
            return customFormat.ifEmpty { defaultFormat }
        }
        set(format) {
            customFormat = format
        }

    val volume: Float
        get() {
            return if (customVolume.isEmpty()) {
                defaultVolume
            } else {
                customVolume.toFloat()
            }
        }

    var header: String
        get() {
            return customHeader.ifEmpty { defaultHeader }
        }
        set(header) {
            customHeader = header
            headerMap.clear()
            customHeader.ifEmpty { defaultHeader }
                .split("\\r\\n")
                .filter { it.isNotEmpty() }
                .associateTo(headerMap) {
                    it.substringBefore(":") to it.substringAfter(":").removePrefix(" ")
                }
        }

    fun setVolume(volume: String) {
        customVolume = volume
    }

    fun clearVideo() {
        video = ""
        time = 0L
    }

    private fun needExtract(isBackGround: Boolean): Boolean {
        val url = url
        if (video.isEmpty()) {
            return true
        }
        if (url.endsWith("m3u8")) {
            return false
        }
        try {
            val current = System.currentTimeMillis() / 1000L
            val timeout = if(isBackGround) 7200L else 60L
            val expire: Long = if (url.startsWith("https://www.youtube.com/")) {
                // Can be "expire=" or "expire/"
                video.substringAfter("expire").substring(1, 11).toLong()
            } else if (url.startsWith("https://hamivideo.hinet.net/") || url.startsWith("https://embed.4gtv.tv/") ||
                url.startsWith("https://www.ftvnews.com.tw/live/live-video/1/")
            ) {
                video.substringAfter("expires").substring(1, 11).toLong()
            } else if (url.startsWith("https://www.ofiii.com/channel/watch/")) {
                time + 14400L
            } else {
                // Assume 2 hours
                time + 7200L
            }
            return current + timeout >= expire
        } catch (e: Exception) {
            Log.e(TAG, "needExtract error", e)
            return true
        }
    }

    suspend fun extract(ytDlp: YtDlp, okHttpClient: OkHttpClient, isBackGround: Boolean = false) {
        mutex.withLock {
            val current = System.currentTimeMillis()
            if (needExtract(isBackGround)) {
                extractMutex(ytDlp, okHttpClient)
                time = current / 1000L
            }
        }
    }

    fun extractMutex(ytDlp: YtDlp, okHttpClient: OkHttpClient) {
        var url = url
        if (url.startsWith("https://hamivideo.hinet.net/") && url.endsWith(".do")) {
            val id = url.substringAfterLast("/").substringBeforeLast(".")
            val okHttpRequestBuilder = Request.Builder()
                .url("https://hamivideo.hinet.net/api/play.do?freeProduct=1&id=$id")
            for (entry in headerMap.entries) {
                okHttpRequestBuilder.addHeader(entry.key, entry.value)
            }
            val okHttpRequest = okHttpRequestBuilder.build()
            okHttpClient.newCall(okHttpRequest).execute().use { response ->
                val jSONObject = JSONObject(response.body.string())
                url = jSONObject.getString("url")
            }
            headerMap.putIfAbsent("Referer", "https://hamivideo.hinet.net/")
        } else if (url.startsWith("https://embed.4gtv.tv/") || url.startsWith("https://www.ftvnews.com.tw/live/live-video/1/")) {
            var id: String
            if (url.startsWith("https://www.ftvnews.com.tw/live/live-video/1/")) {
                id = url.substringAfterLast("/")
            } else {
                val okHttpRequest = Request.Builder().url(url).build()
                okHttpClient.newCall(okHttpRequest).execute().use { response ->
                    val doc = Jsoup.parse(response.body.string())
                    val script = doc.selectFirst("script:containsData(ChannelId)")?.data() ?: throw IOException("找不到 script")
                    id = script.substringAfter("ChannelId: \"").substringBefore("\"")
                }
            }
            val okHttpRequest = Request.Builder()
                .url("https://app.4gtv.tv/Data/GetChannelURL_Mozai.ashx?callback=channelname&Type=LIVE&ChannelId=$id")
                .build()
            okHttpClient.newCall(okHttpRequest).execute().use { response ->
                val bodyString = response.body.string()
                url = bodyString.substringAfter("VideoURL\":\"").substringBefore("\"")
            }
        } else if (url.startsWith("https://www.ofiii.com/channel/watch/")) {
            // Get device id
            val deviceIdRequest = Request.Builder().url("https://www.ofiii.com/api/deviceId").build()
            val deviceId: String
            okHttpClient.newCall(deviceIdRequest).execute().use { response ->
                deviceId = response.body.string().removeSurrounding("\"")
            }
            val id = url.substringAfterLast("/")
            val okHttpRequestBuilder = Request.Builder()
                .url("https://cdi.ofiii.com/ofiii_cdi/video/urls?device_type=pc&device_id=$deviceId&media_type=channel&asset_id=$id&project_num=OFWEB00")
            for (entry in headerMap.entries) {
                okHttpRequestBuilder.addHeader(entry.key, entry.value)
            }
            val okHttpRequest = okHttpRequestBuilder.build()
            okHttpClient.newCall(okHttpRequest).execute().use { response ->
                val jSONObject = JSONObject(response.body.string())
                url = jSONObject.getJSONArray("asset_urls").getString(0)
            }
        }

        val option = Python.getInstance().builtins.callAttr("dict")
        option.callAttr("__setitem__", "format", format)
        if (!headerMap.isEmpty()) {
            val headerDict = Python.getInstance().builtins.callAttr("dict")
            for (entry in headerMap.entries) {
                headerDict.callAttr("__setitem__", entry.key, entry.value)
            }
            option.callAttr("__setitem__", "http_headers", headerDict)
        }
        for (entry in ydlOptions.entries) {
            option.callAttr("__setitem__", entry.key, entry.value)
        }
        video = ytDlp.extract(url, option)
    }

    companion object {
        private const val TAG = "Channel"
    }
}
